cumulus/parachains/runtimes/assets/common/src/matching.rs

```diff
@@ -82,3 +84,163 @@ impl<SelfParaId: Get<ParaId>> ContainsPair<MultiLocation, MultiLocation>
 		}
 	}
 }
+
+/// Adapter verifies if it is allowed to receive `MultiAsset` from `MultiLocation`.
+///
+/// Note: `MultiLocation` has to be from different global consensus.
+pub struct IsTrustedBridgedReserveLocationForConcreteAsset<UniversalLocation, Reserves>(
+	sp_std::marker::PhantomData<(UniversalLocation, Reserves)>,
+);
+impl<
+		UniversalLocation: Get<InteriorMultiLocation>,
+		Reserves: Get<sp_std::vec::Vec<FilteredLocation>>,
+	> ContainsPair<MultiAsset, MultiLocation>
+	for IsTrustedBridgedReserveLocationForConcreteAsset<UniversalLocation, Reserves>
+{
+	fn contains(asset: &MultiAsset, origin: &MultiLocation) -> bool {
+		let universal_source = UniversalLocation::get();
+		log::trace!(
+			target: "xcm::contains",
+			"IsTrustedBridgedReserveLocationForConcreteAsset asset: {:?}, origin: {:?}, universal_source: {:?}",
+			asset, origin, universal_source
+		);
+
+		// check remote origin
+		let _ = match ensure_is_remote(universal_source, *origin) {
+			Ok(devolved) => devolved,
+			Err(_) => {
+				log::trace!(
+					target: "xcm::contains",
+					"IsTrustedBridgedReserveLocationForConcreteAsset origin: {:?} is not remote to the universal_source: {:?}",
+					origin, universal_source
+				);
+				return false
+			},
+		};
+
+		// check asset location
+		let asset_location = match &asset.id {
+			Concrete(location) => location,
+			_ => return false,
+		};
+
+		// check asset according to the configured reserve locations
+		for (reserve_location, asset_filter) in Reserves::get() {
+			if origin.eq(&reserve_location) && asset_filter.matches(asset_location) {
+				return true
+			}
+		}
+
+		false
+	}
+}
+
+/// Disallow all assets the are either not `Concrete`, or not explicitly allowed by
+/// `LocationAssetFilters`, iff `dest` matches any location in `LocationAssetFilters`.
+///
+/// Returns `false` regardless of `assets`, if `dest` does not match any location in
+/// `LocationAssetFilters`. Otherwise, returns `true` if asset is either not `Concrete` or is not
+/// explicitly allowed by `LocationAssetFilters`, otherwise returns `false`.
+pub struct DisallowConcreteAssetUnless<LocationAssetFilters>(
```

Comment:
@serban300 well, so what do you suggest here? 
Adrian changed this here:
https://github.com/paritytech/polkadot-sdk/pull/1352#discussion_r1315000670
https://github.com/paritytech/polkadot-sdk/pull/1352/commits/867321c2d2ef7480ff12c762ad5e876924b73759

Reply:
Starting from `type XcmReserveTransferFilter = EverythingBut<bridging::DisallowForReserveTransfer>;` I think something like `XcmReserveTransferFilter = bridging::ReserveBasedTransferAssets` would be easier to understand.

Then for
```
	pub type DisallowForReserveTransfer = ExcludeOnlyForRemoteDestination<
		UniversalLocation,
		(to_wococo::FilteredNetworkExportTable, to_rococo::FilteredNetworkExportTable),
		(
			DisallowConcreteAssetUnless<to_wococo::AllowedReserveTransferAssetsLocations>,
			DisallowConcreteAssetUnless<to_rococo::AllowedReserveTransferAssetsLocations>,
		),
	>;
```

I would replace it with something like
```
	pub type ReserveBasedTransferAssets = AssetsMatchingDestinationAndFilter<
		UniversalLocation,
		(to_wococo::FilteredNetworkExportTable, to_rococo::FilteredNetworkExportTable),
		(
			to_wococo::AllowedReserveTransferAssetsLocations,
			to_rococo::AllowedReserveTransferAssetsLocations,
		),
	>;
```

Something like this, if it makes sense, just as a general idea. Such that each filter describes what it allows, not what it denies. Didn't actually try it. Just threw some names around.

Reply:
yes, I agree we can hide away some of the adapters and have clearer helper types:

`EverythingBut<bridging::DisallowForReserveTransfer>` -> `bridging::ReserveBasedTransferAssets` and so on - hide away some of the complicated logic - (everything_but(disallow_unless(allowed_locations)) is hard to follow

@serban300 maybe you can have a try at it since you came in with fresh eyes and (yet) unbiased view 😆

Reply:
Sure, I can give it a try.

Reply:
Or we can tackle this in a follow-up PR

Reply:
I found another easier way without `EveryrhingBut` - just to list supported cases, monday will be fixed

Reply:
@serban300 ok, [here](https://github.com/paritytech/polkadot-sdk/pull/1215/commits/ea0997be227389a22041f0c6c4e4446f73a22298) it is fixed, all that mind-bending stuff `EverythingBut+ExcludeOnlyForRemoteDestination+DisallowConcreteAssetUnless` is removed, now we just explicitly list all cases e.g.
- to dest with `parents: 0` and `parents: 1` allow for all assets (like was before)
- for bridging, we just say that for dest AssetHub allow only particular assets, e.g. to the remote AssetHubKusama allow to send only DOTs

Reply:
Thank you ! Looks good !